[Taoseeker] iP

Taoseeker opened the pull request:
# NeoChat
>“Smart people have already started using honorifics when talking to AI.”  - Me

Neochat frees your mind of having to remember things you need to do. 
It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

All you need to do is

1. download it from [here](https://github.com/Taoseeker/ip/releases/tag/A-Jar)
2. Turn on intelliJ idea
3. make sure the `.jar` file is in working directory
4. in your terminal, run the command `java -jar "neochat.jar"`
5. add your tasks
6. let it manage your tasks for you 😉

And it's **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events
- [ ] AI assistant (coming sooooooon)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Taoseeker committed:
Renamed chatbot and added a skeletal structure

Taoseeker committed:
Added echo feature

Taoseeker committed:
Added a todo list

Taoseeker committed:
Implement mark a todo task as done and a class Task to represent tasks.

Taoseeker committed:
Added 3 different types of tasks and use inheritence to handle the different behavior of different tasks. Handle certain exceptions like invalid input.

Taoseeker committed:
Implemented  automated text UI testing.

Taoseeker committed:
Added new exceptions and logics to handle invalid input by the user.

Taoseeker committed:
Added delete command. Edited some text.

Taoseeker committed:
put everything into one dummy package neochat

Taoseeker committed:
Add a method toFileString to write task in to save file.

Taoseeker committed:
Add new feature: save tasklist on local disk.

Taoseeker committed:
Merge branch 'master' of https://github.com/Taoseeker/ip

Taoseeker committed:
Edited Task such that the time become LocalDateTime instead of String

Taoseeker committed:
Modify the behavior of time and date. Add a parser class to intepret time and date

Taoseeker committed:
Merge branch 'branch-Level-8'

Taoseeker committed:
Modifiy behavoir of time and date

Taoseeker committed:
Make the code more OOP by extracting classes like Ui, Parser

Taoseeker committed:
Organize different classes into multiple packages

Taoseeker committed:
Merge branch 'add-gradle-support' of https://github.com/Taoseeker/ip

Taoseeker committed:
Update .gitignore

Taoseeker committed:
Incorporated Gradle in NeoChat

Taoseeker committed:
Add JUnit test. Add mockito dependency to Gradle.

Taoseeker committed:
Release NeoChat v0.1

Taoseeker committed:
Add JavaDoc to majority of the classes and public methods and some private methods.

Taoseeker committed:
Modify code to comply with suggested coding style.

Taoseeker committed:
Add new feature: find tasks in task list by keywords

Taoseeker committed:
Modified .gitignore to untrack redundant file

Taoseeker committed:
Merge branch 'branch-Level-9'

Taoseeker committed:
Merge branch 'A-CodingStandard'

# Conflicts:
#	src/main/java/neochat/task/Parser.java

Resolution: keep both

Taoseeker committed:
Merge branch 'A-JavaDoc'

# Conflicts:
#	src/main/java/neochat/task/Parser.java
#	src/main/java/neochat/task/tasklist/TaskList.java

Resolution: keep both

Nano-233 commented on src/main/java/neochat/task/Parser.java:
```diff
@@ -0,0 +1,152 @@
+                System.out.println("Unknown command type" + commandType);
+            }
+        } catch (IllegalArgumentException e) {
+            e.printStackTrace();
```
would a error log or user-friendly message be better than printing the internal details? for example e.getmessage

Taoseeker replied:
Thank you for your suggestion! I was thinking using this could help me debug. Will make the error message more user-friendly in future release.

Nano-233 commented on src/main/java/neochat/task/Parser.java:
```diff
@@ -0,0 +1,152 @@
+    private void parseDeadline(String input) {
+        String[] parts = input.split(" /by ", 2);
+        if (parts.length < 2) {
+            throw new IllegalArgumentException("Deadline format error，the correct format should be" +
```
Is there a reason that a unstandardized comma was used here instead of the standard comma? I noticed it in a few other places too

Taoseeker replied:
No specific reason, it is just my standard keyboard input is Chinese(I wrote short comment in Chinese sometime for my own reference) so sometimes I mistakenly use Chinese comma.

Nano-233 commented on src/main/java/neochat/task/Parser.java:
```diff
@@ -0,0 +1,152 @@
+        }
+    }
+
+    private static void printCommandList() {
+        System.out.println("____________________________________________________________");
+        System.out.println("Here are the available commands:");
+        System.out.println("1. list - Show all tasks");
+        System.out.println("2. todo <description> - Add a Todo task");
+        System.out.println("3. deadline <description> /by <yyyy-MM-dd HH:mm> - Add a Deadline task");
+        System.out.println("4. event <description> /from <yyyy-MM-dd HH:mm> /to <yyyy-MM-dd HH:mm> - Add an Event task");
+        System.out.println("5. mark <task number> - Mark a task as done");
+        System.out.println("6. unmark <task number> - Mark a task as not done yet");
+        System.out.println("7. help - Show the command list");
+        System.out.println("8. bye - Exit the program");
+        System.out.println("____________________________________________________________");
+    }
```
Perhaps a separate utility class is most suitable for this static method as it is unrelated to parsing?

Taoseeker replied:
You do have a valid point. I plan to make Ui class handle this in future release.

Nano-233 commented on src/main/java/neochat/task/tasklist/TaskList.java:
```diff
@@ -0,0 +1,295 @@
+package neochat.task.tasklist;
+
+import java.io.*;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.*;
+
+import neochat.task.*;
```
Perhaps non-wildcard imports would be more suitable for our coding standards? I assume IDE issue since other ones are fine?

Taoseeker replied:
Thanks for pointing it out. Will use explicit import in future.

Nano-233 commented on src/main/java/neochat/task/tasklist/TaskList.java:
```diff
@@ -0,0 +1,295 @@
+     */
+    public TaskList() {
+        this.tasks = new ArrayList<>(100);
+        this.savedListFile = new File("src/data/savedList.txt");
```
Perhaps using a constant could be more suitable for file handling to allow for easier modifications and prevent hard-coded path related errors?

Nano-233 commented on src/main/java/neochat/task/tasklist/TaskList.java:
```diff
@@ -0,0 +1,295 @@
+    private Task parseTask(String line) {
+        String[] parts = line.split(" \\| ");
+        if (parts.length < 3) {
+            return null;
```
perhaps a log of invalid format would be useful for debugging (as there is an invalid line)

Taoseeker replied:
You are right: my initial thought is to let caller of `parseTask` to handle null, but I forgot to implement it. Will do so in future version.

Nano-233 commented on src/main/java/neochat/task/tasklist/TaskList.java:
```diff
@@ -0,0 +1,295 @@
+     * Displays all tasks in the list with their indexes.
+     * Shows "Empty task list!" if no tasks exist.
+     */
+    public void printList() {
+        if (count == 0) {
+            System.out.println("Empty task list!");
+        } else {
+            for (int i = 0; i < count; i++) {
+                Task task = tasks.get(i);
+                System.out.println((i + 1) + ": " + task.toString());
+            }
+        }
```
is there a reason a global count is used instead of for example task.size() ?

Taoseeker replied:
I wrote my own data structure in an early version which does not keep track of the count, so this global count is 'inherited' from earlier version and I forgot to edit.

Nano-233 commented on src/main/java/neochat/task/tasklist/TaskList.java:
```diff
@@ -0,0 +1,295 @@
+                throw new IndexOutOfBoundsException();
+            }
+            Task task = tasks.get(taskIndex);
+            task.markDone();
+            System.out.println("____________________________________________________________");
+            System.out.println("Nice! I've marked this task as done:");
+            System.out.println("  " + task);
+            System.out.println("____________________________________________________________");
+        } catch (NumberFormatException e) {
+            System.out.println("Invalid input. Please provide a valid task number.");
+        } catch (IndexOutOfBoundsException e) {
+            System.out.println("Invalid task number. Please provide a number between 1 and " + count + ".");
+        }
+    }
+
+    /**
+     * Marks a task as not done based on its position in the list.
+     *
+     * @param input String containing the task number (1-based index)
+     * @throws NumberFormatException If input is not a valid integer
+     * @throws IndexOutOfBoundsException If task number is out of valid range
+     */
+    public void markAsNotDone(String input) {
+        try {
+            int taskIndex = Integer.parseInt(input) - 1;
+            if (taskIndex < 0 || taskIndex >= count) {
+                throw new IndexOutOfBoundsException();
+            }
+            Task task = tasks.get(taskIndex);
+            task.markNotDone();
+            System.out.println("____________________________________________________________");
+            System.out.println("OK, I've marked this task as not done yet:");
+            System.out.println("  " + task);
+            System.out.println("____________________________________________________________");
+        } catch (NumberFormatException e) {
+            System.out.println("Invalid input. Please provide a valid task number.");
+        } catch (IndexOutOfBoundsException e) {
+            System.out.println("Invalid task number. Please provide a number between 1 and " + count + ".");
+        }
+    }
```
Perhaps could combine the two methods as they seem to share many lines of code?

Taoseeker replied:
I will consider it.

Nano-233 reviewed (commented):
LGTM! Just small alternatives that perhaps could be considered to make the code easier for maintaining and reading!

weiwong834 commented on src/main/java/neochat/task/tasklist/TaskList.java:
```diff
@@ -0,0 +1,295 @@
+        try {
+            return LocalDateTime.parse(dateTimeStr, DATE_TIME_FORMATTER);
+        } catch (DateTimeParseException e) {
+            throw new IllegalArgumentException("Invalid time format, the correct format should be "
```
Would a checked exception such as DateTimeFormatException be better in this case?

weiwong834 commented on src/main/java/neochat/task/tasklist/TaskList.java:
```diff
@@ -0,0 +1,295 @@
+        } catch (IndexOutOfBoundsException e) {
+            System.out.println("Invalid task number. Please provide a number between 1 and " + count + ".");
+        }
+    }
```
Maybe the all the println lines can be handled in the Ui class to simplify these methods.

weiwong834 commented on src/main/java/neochat/ui/Ui.java:
```diff
@@ -0,0 +1,20 @@
+package neochat.ui;
+
+import java.util.*;
```
Maybe consider not using wildcard imports

weiwong834 reviewed (commented):
Overall looks good! Perhaps a shifting some of the implementations into Ui class will improve clarity.

Taoseeker committed:
Add GUI

Taoseeker committed:
Integrate assertions

Taoseeker committed:
Improve code quality

There are several methods names that are unintuitve. Change the names into more intuitive ones.

Some methods are not self-explanatory. Add comment to these methods to improve readability.

Delete some redundant comments.

Taoseeker committed:
Integrate assertions

Add assertions in during the parse and interpretion of user input.
Add assertions in during the parse and interpretion of saved data.

Taoseeker committed:
Merge pull request #1 from Taoseeker/A-CodeQuality

Improve code quality

Taoseeker committed:
Merge branch 'A-Assertions'

Taoseeker committed:
Add new feature: archive and loadarchive

Taoseeker committed:
Remove dead code

Taoseeker committed:
Update Junit test

Taoseeker committed:
Update JUnit Test

Taoseeker committed:
Remove dead code

Taoseeker committed:
Remove dead code

Taoseeker committed:
Update JavaDoc

Taoseeker committed:
Update help information

Taoseeker committed:
Refine Gui

Taoseeker committed:
Refine Gui

Taoseeker committed:
Update Gui screenshot and README.md

Taoseeker committed:
Refine code style

Taoseeker committed:
Update user guide

Taoseeker committed:
Update user guide

Taoseeker committed:
Update README.md

Taoseeker committed:
Update README.md
